File: python/ByteClient.py
#!/usr/bin/env python
# encoding: utf-8

import zmq

class ByteClient(object):
  def __init__(self, sock_name):
      self._ctx = zmq.Context()
      self.pipe = self._ctx.socket(zmq.REQ)
      self.pipe.connect(sock_name)

  def send_byte(self, string):
      self.pipe.send_string(string)

  def recv_byte(self):
      return self.pipe.recv_string()

  def send_and_recv(self, string):
      self.send_byte(string)
      return self.recv_byte()

if __name__ == '__main__':
  bc = ByteClient(r"ipc:///tmp/sock_temp")
  try:
    while (True):
      s = input()
      print(bc.send_and_recv(s))
  except KeyboardInterrupt:
    print("bye.")

File: cs/ByteComm/ByteClient.cs
using System;
using System.Text;
using NetMQ;
using NetMQ.Sockets;

namespace ByteComm
{
    public class ByteClient
    {
        readonly RequestSocket socket;

        public ByteClient(string sockname)
        {
            socket = new RequestSocket(sockname); 
        }

        public void Send_Byte(byte[] message)
        {
            socket.SendFrame(message);
        }

        public byte[] Recv_Byte()
        {
            return socket.ReceiveFrameBytes();
        }

        public void Send_String(string message)
        {
            socket.SendFrame(Encoding.ASCII.GetBytes(message));
        }

        public string Recv_String()
        {
            return socket.ReceiveFrameString();
        }

        public byte[] Send_And_Recv(byte[] message)
        {
            Send_Byte(message);
            return Recv_Byte();
        }

        public string Send_And_Recv(string message)
        {
            Send_String(message);
            return Recv_String();
        }

        public void Close()
        {
            socket.Close();
        }
    }
}
